### PR TITLE
[IMP] delivery: Improve inheritability of price computation

### PR DESCRIPTION
**Pull request by victoralmau**
**Description of the issue/feature this PR addresses**:
With this commit, third party modules can extend delivery price rules for:

- Adding extra rule criteria.
- Adding extra variable factor data for computation.
- Allowing to modify existing criteria/variable factor.

An example of an extra criteria can be the "volumetric weight", which is used for some carriers to compute the price for light packages, but very big in volume.

**Impacted versions**:
- master

Related to: https://github.com/odoo/odoo/pull/74561#issuecomment-890977554

cc @Tecnativa TT30670

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/delivery/models/delivery_grid.py [hunk 1]
```diff
@@ -90,24 +90,25 @@ def _get_price_available(self, order):
 
         total = order.currency_id._convert(
             total, order.company_id.currency_id, order.company_id, order.date_order or fields.Date.today())
+        params = {"price": total, "volume": volume, "weight": weight, "quantity": quantity}
+        params.update(extra_params)
```
Comment (pedrobaeza) [inline addons/delivery/models/delivery_grid.py:94]:
```suggestion
        if extra_params:
            params.update(extra_params)
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
First line of commit message should be short and explicit, and the rest must explain the why and more details of the what, but not repeating what can be seen in the code:
```
[IMP] delivery: Improve inheritability of price computation

With this commit, third party modules can extend delivery price rules for:

- Adding extra rule criteria.
- Adding extra variable factor data for computation.
- Allowing to modify existing criteria/variable factor.

An example of an extra criteria can be the "volumetric weight", which is used for some carriers to compute the price for light packages, but very big in volume.
```

**Comment by pedrobaeza** [comment]
Check also why runbot is red

File: addons/delivery/models/delivery_grid.py [hunk 2]
```diff
@@ -90,25 +90,27 @@ def _get_price_available(self, order):
...
         }
 
-    def _get_price_from_picking(self, total, weight, volume, quantity):
+    def _get_price_from_picking(self, params):
         price = 0.0
         criteria_found = False
-        price_dict = self._get_price_dict(total, weight, volume, quantity)
-        if self.free_over and total >= self.amount:
+        price_dict = self._get_price_dict(params)
+        if self.free_over and price_dict["price"] >= self.amount:
```
Comment (pedrobaeza) [inline addons/delivery/models/delivery_grid.py:113]: As we are avoiding bad dictionary param all time, let's do the same here.

```suggestion
        if self.free_over and price_dict.get("price", 0) >= self.amount:
```

**Review by pedrobaeza — APPROVED** [review]

**Comment by C3POdoo** [comment]
Dear @victoralmau,

Thank you for your contribution but we are closing it due to inactivity.
We only support the last 3 stable versions so no longer accepts patches into this branch.

We apology if we could not look at your request in time.
If the contribution still makes sense, please let us know and do not hesitate to recreate one for the recent versions. We will try to check it as soon as possible.

_This is an automated message._

**Comment by victoralmau** [comment]
I don't quite understand the automatic message in this case (it is done in master). Can it be reopened or is it better to create a new PR at 17.0?

**Pull request closed without merging**